src/opencl/opencl.cpp

```diff
@@ -216,130 +329,276 @@ void sloth256_189_opencl_batch_encode_instance(unsigned char* inout_offset, size
+                                         1, &offset, &global_size, 
+                                         &local_size, 0, NULL, NULL); 
+            if (err != CL_SUCCESS) return;
         }
-        //printf("Elapsed time for encode kernel: %0.f milliseconds\n", duration / 1000000);
 
+        // Copy the piece data back to the host in multiple iterations
         for (size_t i = 0; i < num_command_queues; i++) {
-            err = clEnqueueReadBuffer(instance->cq[i], piecebuf, CL_FALSE, i * divided_size, divided_size, (void*)(inout_offset + divided_size * i + copy_offset), 0, NULL, NULL); if (err != CL_SUCCESS) return;
+            err = clEnqueueReadBuffer(instance->cq[i], piecebuf, CL_FALSE,
+                                      i * divided_size, divided_size,
+                                      (void*)(inout_offset + divided_size * i + copy_offset),
+                                      0, NULL, NULL);
+            if (err != CL_SUCCESS) return;
         }
 
-        err = clReleaseMemObject(piecebuf); if (err != CL_SUCCESS) return;
-        err = clReleaseMemObject(ivbuf); if (err != CL_SUCCESS) return;
+        // Free the buffers allocated for IVs and pieces
+        err = clReleaseMemObject(piecebuf); 
+        if (err != CL_SUCCESS) return;
+
+        err = clReleaseMemObject(ivbuf);
+        if (err != CL_SUCCESS) return;
 
         remaining_size -= next_size;
     }
 
-    err = clFlush(instance->cq[num_command_queues - 1]); if (err != CL_SUCCESS) return;
-    err = clFinish(instance->cq[num_command_queues - 1]); if (err != CL_SUCCESS) return;
+    err = clFlush(instance->cq[num_command_queues - 1]); 
+    if (err != CL_SUCCESS) return;
 
-    //auto end = std::chrono::high_resolution_clock::now();
-    //std::cout << "Total elapsed time for GPU encode on platform " << (int)instance->platform << ": " << (std::chrono::duration<double>(end - start)).count() << " seconds" << std::endl;
+    err = clFinish(instance->cq[num_command_queues - 1]);
+    if (err != CL_SUCCESS) return;
+
+    /*auto end = std::chrono::high_resolution_clock::now();
+    std::cout << "Total elapsed time for GPU encode on platform " 
+              << (int)instance->platform << ": " 
+              << (std::chrono::duration<double>(end - start)).count() 
+              << " seconds" << std::endl;*/
 }
 
+// Values that determine how much of the work is delegated to the respective platforms
+// For example, on a system where there are only Nvidia GPUs and CPUs, and the CPU and NVIDIA
+// factors are 80.0f and 20.0f respectively, 80% of the data will be encoded by the Nvidia GPUs
+// and 20% of the data will be encoded by the CPUs
+//
+// Currently we hand-tune the factors for each system. The commented out codes above time the
+// execution of the encode operation on each platform to give us a hint on how to tune the factors
 #define CPU_FACTOR 0.75f
 #define NVIDIA_FACTOR 100.0f
 #define AMD_FACTOR 15.0f
 #define INTEL_FACTOR 1.3f

```

Comment:
It would be really-really nice to have some kind of function that does quick benchmark and returns these constants in form of a struct that can be later used as an argument for `sloth256_189_opencl_batch_encode`. This is something that will be harder for us to implement on our own.